File: src/spi.c
#include <errno.h>
#include <fcntl.h>
#include <linux/spi/spi.h>
#include <linux/spi/spidev.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include "spi.h"

void spi_init(spi_t* self, const char* dev_path) {
  // open the SPI device
  self->fd = open(dev_path, O_RDWR);
  if (self->fd < 0) {
    fprintf(stderr, "couldn't open SPI device %s\n", dev_path);
    exit(1);
  }

  // SPI config
  uint8_t buf8 = 0;  // CPOL=0, CPHA=0
  IOCTL_WITH_ERR_HANDLING(self->fd, SPI_IOC_WR_MODE, &buf8,
                          "couldn't set SPI mode\n");
  buf8 = 0;  // MSB first
  IOCTL_WITH_ERR_HANDLING(self->fd, SPI_IOC_WR_LSB_FIRST, &buf8,
                          "couldn't set SPI MSB/LSB first\n");
  buf8 = 8;
  IOCTL_WITH_ERR_HANDLING(self->fd, SPI_IOC_WR_BITS_PER_WORD, &buf8,
                          "couldn't set SPI bit count per word\n");
  uint32_t buf32 = 100000;
  IOCTL_WITH_ERR_HANDLING(self->fd, SPI_IOC_WR_MAX_SPEED_HZ, &buf32,
                          "couldn't set SPI max speed\n");
}

void spi_deinit(spi_t* self) {
  // close the SPI device
  close(self->fd);
}

void spi_write_reg8(spi_t* self, uint8_t addr, uint8_t val) {
  struct spi_ioc_transfer xfer[1];
  memset(xfer, 0, sizeof(xfer));
  uint8_t buf[2] = {addr, val};
  buf[0] |= 0x80;  // set the address MSB to indicate write, as opposed to read
  xfer[0].tx_buf = (uint64_t)buf;
  xfer[0].rx_buf = 0;
  xfer[0].len = 2;

  IOCTL_WITH_ERR_HANDLING(self->fd, SPI_IOC_MESSAGE(1), xfer,
                          "couldn't perform SPI register write");
}

void spi_write_bulk(spi_t* self, uint8_t addr, uint8_t* data, uint8_t len) {
  struct spi_ioc_transfer xfer[1];
  memset(xfer, 0, sizeof(xfer));
  uint8_t* buffer = malloc((size_t)(len + 1));
  buffer[0] = addr | 0x80;
  memcpy(buffer + 1, data, (size_t)len);
  xfer[0].tx_buf = (uint64_t)buffer;
  xfer[0].rx_buf = 0;
  xfer[0].len = len + 1;

  IOCTL_WITH_ERR_HANDLING(self->fd, SPI_IOC_MESSAGE(1), xfer,
                          "couldn't perform SPI register write");

  free(buffer);
}

uint8_t spi_read_reg8(spi_t* self, uint8_t addr) {
  struct spi_ioc_transfer xfer[2];
  memset(xfer, 0, sizeof(xfer));
  uint8_t val;
  // don't set the address MSB to indicate read, as opposed to write
  xfer[0].tx_buf = (uint64_t)(&addr);
  xfer[0].rx_buf = 0;
  xfer[0].len = 1;
  xfer[1].tx_buf = 0;
  xfer[1].rx_buf = (uint64_t)(&val);
  xfer[1].len = 1;

  IOCTL_WITH_ERR_HANDLING(self->fd, SPI_IOC_MESSAGE(2), xfer,
                          "couldn't perform SPI register read");

  return val;
}

File: src/spi.zig
const std = @import("std");
const c = @import("c");
const misc = @import("misc.zig");

pub const Spi = struct {
    dev_file: std.Io.File,

    const Logger = std.log.scoped(.spi);

    pub fn init(io: std.Io, dev_path: []const u8) !Spi {
        // const fd: i32 = @intCast(std.os.linux.open(dev_path, .{}, c.O_RDWR));
        const dev_file = try std.Io.Dir.openFileAbsolute(io, dev_path, .{ .mode = .read_write });
        const fd = dev_file.handle;

        var buf8: u8 = 0; // CPOL=0, CPHA=0
        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_WR_MODE,
            @intFromPtr(&buf8),
            "couldn't set SPI mode",
        );
        buf8 = 0; // MSB first
        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_WR_LSB_FIRST,
            @intFromPtr(&buf8),
            "couldn't set SPI MSB/LSB first",
        );
        buf8 = 8;
        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_WR_BITS_PER_WORD,
            @intFromPtr(&buf8),
            "couldn't set bit count per word",
        );
        var buf32: u32 = 100000;
        // var buf32: u32 = 1000;
        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_WR_MAX_SPEED_HZ,
            @intFromPtr(&buf32),
            "couldn't set max speed",
        );

        Logger.info("SPI ready", .{});

        return .{
            .dev_file = dev_file,
        };
    }

    pub fn deinit(self: *const Spi, io: std.Io) void {
        self.dev_file.close(io);
    }

    pub fn write_reg8(self: *Spi, addr: u8, val: u8) !void {
        const fd = self.dev_file.handle;
        var xfer = std.mem.zeroes(c.spi_ioc_transfer);
        var buf = [2]u8{ addr, val };
        buf[0] |= 0x80; // Indicate write
        xfer.tx_buf = @intFromPtr(&buf);
        xfer.rx_buf = 0;
        xfer.len = 2;

        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_MESSAGE_1,
            @intFromPtr(&xfer),
            "couldn't perform SPI register write",
        );
    }

    pub fn write_bulk(self: *Spi, addr: u8, data: []const u8, gpa: std.mem.Allocator) !void {
        const fd = self.dev_file.handle;
        var xfer = std.mem.zeroes(c.spi_ioc_transfer);
        const buf = try gpa.alloc(u8, data.len + 1);
        defer gpa.free(buf);
        buf[0] = addr | 0x80; // Indicate write
        @memcpy(buf[1..], data);
        xfer.tx_buf = @intFromPtr(buf.ptr);
        xfer.rx_buf = 0;
        xfer.len = data.len + 1;

        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_MESSAGE_1,
            @intFromPtr(&xfer),
            "couldn't perform SPI register write",
        );
    }

    pub fn read_reg8(self: *Spi, addr: u8) !u8 {
        const fd = self.dev_file.handle;
        var xfer = std.mem.zeroes([2]c.spi_ioc_trasnfer);
        var val: u8 = undefined;
        xfer[0].tx_buf = @intFromPtr(&addr);
        xfer[0].rx_buf = 0;
        xfer[0].len = 1;
        xfer[1].tx_buf = 0;
        xfer[1].rx_buf = @intFromPtr(&val);
        xfer[1].len = 1;

        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_MESSAGE_2,
            @intFromPtr(&xfer),
            "coulnd't perform SPI register read",
        );
        return val;
    }

    pub fn read_bulk(self: *Spi, addr: u8, result: []u8) !void {
        const fd = self.dev_file.handle;
        var xfer = std.mem.zeroes([2]c.spi_ioc_trasnfer);
        xfer[0].tx_buf = @intFromPtr(&addr);
        xfer[0].rx_buf = 0;
        xfer[0].len = 1;
        xfer[1].tx_buf = 0;
        xfer[1].rx_buf = @intFromPtr(result.ptr);
        xfer[1].len = result.len;

        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_MESSAGE_2,
            @intFromPtr(&xfer),
            "coulnd't perform SPI register read",
        );
    }

    pub fn transact(self: *Spi, data_tx: []const u8, data_rx: []u8) !void {
        if (data_tx.len != data_rx.len) {
            return error.LengthMismatch;
        }

        const fd = self.dev_file.handle;
        var xfer = std.mem.zeroes(c.spi_ioc_transfer);
        xfer.tx_buf = @intFromPtr(data_tx.ptr);
        xfer.rx_buf = @intFromPtr(data_rx.ptr);
        xfer.len = @intCast(data_tx.len);

        _ = try misc.ioctl_checked(
            fd,
            c.SPI_IOC_MESSAGE_1,
            @intFromPtr(&xfer),
            "coulnd't perform SPI transaction",
        );

        Logger.debug("SPI transaction: tx={any}, rx={any}", .{
            data_tx,
            data_rx,
        });
    }
};
